File: bucket_sort/Clang/bucketSort.c
// Bucket Sort in C

#include <stdio.h>

int getMax(int array[], int size)
{
    int max = array[0];
    for (int i = 1; i < size; i++)
        if (array[i] > max)
            max = array[i];
    return max;
}

void bucketSort(int array[], int size)
{
    int bucket[10];
    const int max = getMax(array, size);
    for (int i = 0; i <= max; i++)
    {
        bucket[i] = 0;
    }
    for (int i = 0; i < size; i++)
    {
        bucket[array[i]]++;
    }
    for (int i = 0, j = 0; i <= max; i++)
    {
        while (bucket[i] > 0)
        {
            array[j++] = i;
            bucket[i]--;
        }
    }
}

void printArray(int array[], int size)
{
    for (int i = 0; i < size; ++i)
    {
        printf("%d  ", array[i]);
    }
    printf("\n");
}

int main()
{
    int data[] = {5, 4, 3, 2, 1};
    int size = sizeof(data) / sizeof(data[0]);
    bucketSort(data, size);
    printf("Sorted array is: \n");
    printArray(data, size);
}

File: bucket_sort/Python/bucketSort.py
# Prg of BUCKET SORT in Python


def insertionSort(b):
	for i in range(1, len(b)):
		up = b[i]
		j = i - 1
		while j >= 0 and b[j] > up:
			b[j + 1] = b[j]
			j -= 1
		b[j + 1] = up
	return b


def bucketSort(x):
	arr = []
	slot_num = 10  # 10 here refers 10 slots; each slot's size being 0.1
	for i in range(slot_num):
		arr.append([])

	# Putting array elements in different buckets
	for j in x:
		index_b = int(slot_num * j)
		arr[index_b].append(j)

	# Sorting individual buckets
	for i in range(slot_num):
		arr[i] = insertionSort(arr[i])

	# Concatenating the result
	k = 0
	for i in range(slot_num):
		for j in range(len(arr[i])):
			x[k] = arr[i][j]
			k += 1
	return x


# MAIN
x = [0.5, 0.4, 0.3, 0.2, 0.1]
print("Sorted Array is :")
print(bucketSort(x))